feat: Hetzner install - step 1

alexgarel opened the pull request:
Proxmox nodes install

alexgarel committed:
fix: fixing variable undefined in motd

to avoid ansible.errors.AnsibleUndefinedVariable: 'public_ipv6' is undefined

alexgarel committed:
docs: how to prepare host

alexgarel committed:
feat: first hetzner install

alexgarel committed:
feat: wip on nftables + fail2ban

alexgarel committed:
refactor: separate external roles

alexgarel committed:
feat: tentative addition of nftable
but this is not compatible with proxmox servers !

alexgarel committed:
Revert "feat: tentative addition of nftable"

This reverts commit e3b05a92214b21d703e1774f9bc1b12179fda88c.

alexgarel committed:
feat: add ipv6 to /etc/hosts

alexgarel committed:
feat: add some base packages

alexgarel committed:
refactor: hide some emails

alexgarel committed:
feat(wip): work on proxmox node install

alexgarel committed:
feat: wip on proxmox + zfs install

alexgarel committed:
feat: wip on proxmox node, now using parted and replicated for other servers

alexgarel committed:
feat: firewall rules

alexgarel committed:
feat: creating users having ssh access

alexgarel committed:
fix: switch to iptables for fail2ban and use right backend

alexgarel committed:
docs: explain passwords decrypt

teleboas commented on roles/sshd/tasks/create_user.yml:
```diff
@@ -0,0 +1,82 @@
+        force: true
+        mode: 0600
+
+    - name: Save password in a file
```
Temporary plaintext files containing passwords represent a potential vulnerability.

Safer alternatives:

  1. Use a pipe or memory-based approach: Rather than writing to a file, the password
  could be piped directly to the encryption command:
  ```
echo "{{ user_password.stdout }}" | openssl pkeyutl -encrypt -inkey /tmp/{{ user.name }}.pem -pubin  -pubin -in /dev/stdin -out {{ user_password_dir }}/{{ ansible_hostname }}-password.txt
```

  2. Use a secure temporary filesystem: Utilize an in-memory filesystem like /dev/shm for
  temporary files.

alexgarel replied:
ok we don't use it anymore.

teleboas commented on roles/sshd/tasks/main.yml:
The current name of this role doesn't quite capture its function, and it's become more problematic now that it's creating separate user accounts.

It might be helpful if we renamed it to something more descriptive. How about `user-management` or `system-users`? These options seem to better reflect the role's purpose.

alexgarel replied:
+1 for system-users

alexgarel replied:
I have separated the role in two. I have kept sshd but added system-users.

teleboas commented on roles/sshd/tasks/create_user.yml:
```diff
@@ -0,0 +1,82 @@
+        mode: 0600
+      register: user_password_file
+
+    - name: encrypt file with user public ssh key
```
If you attempt to use an Ed25519 or ECDSA key with `openssl pkeyutl -encrypt`, you'll get the following error:
```
Error: Public key operation not supported for this keytype
```

This is a significant limitation in the current implementation. If a user has an Ed25519 key on GitHub (which is increasingly common as Ed25519 is recommended for SSH keys due to its security and performance), the password encryption process  would fail.

Given the rising popularity of Ed25519 keys for SSH authentication, this issue could affect many users.

alexgarel replied:
Well spotted, even if it makes me a bit sad :sweat: 

Ed25519 encryption is indeed not possible (there seems to be ways to derive them to encryption key, but it's not yet well established software).

Alternatives I though of:
* send password by email
* store password hash (in a git-crypt file) and reuse it on every system
* don't handle it, and handle it manually (but I see it as a lot of burden)

teleboas replied:
How about letting users set up their own passwords? They can already connect using their public keys.
We would not be able to store it in the repository, but do we really need it there?

alexgarel replied:
Ok, I was not aware of this method (`passwd -e -d <user-name>`) and it's perfectly fine indeed.

alexgarel replied:
done.

teleboas commented on roles/sshd/tasks/main.yml:
```diff
@@ -20,12 +20,26 @@
+  ansible.builtin.include_tasks: create_user.yml
+  vars:
+    user: "{{ item }}"
+    user_has_password: "{{ item.is_superuser and is_proxmox_host }}"
```
This line feels like a violation of the separation of concerns principle:

This creates a direct dependency between the user management role (for now named `sshd`) and Proxmox-specific logic, which is problematic for several reasons:

  1. Role Reusability: This role should be usable in any context, not just with Proxmox hosts
  2. Single Responsibility: The role should focus solely on managing users without knowledge of specific infrastructure platforms
  3. Maintenance Challenges: Changes to Proxmox-related logic might require modifications to this user management role
  4. Testing Complexity: Testing the role in isolation becomes difficult due to the Proxmox dependency

  A better approach could be:

  1. Pass password requirements as a parameter to the role from the playbook level, where infrastructure decisions belong:
 ```
  # In the playbook
  - name: Include user management role
    include_role:
      name: sshd
    vars:
      enable_user_passwords: "{{ is_proxmox_host }}"
```

teleboas commented on group_vars/all/common.yml:
```diff
@@ -0,0 +1,38 @@
+off_internal_hosts_ips_v4:
```
This list should ideally be generated from the inventory. 
This would:
  1. Avoid manual maintenance
  2. Ensure consistency
  3. Prevent errors when servers change

Here is one way to do it:
```
# Generate off_internal_hosts_ips_v4 dynamically from inventory
off_internal_hosts_ips_v4: "{{ groups['all'] | map('extract', hostvars, ['ansible_ssh_host'])
| list | reject('undefined') }}"

# Additional IPs not in inventory (they should probably be added to the inventory)
off_additional_ips_v4:
  - "213.36.253.215"  # off1 secondary
  - "213.36.253.214"  # off2 secondary
  - "46.105.72.120"   # ovh1 secondary
  - "45.147.209.254"  # moji

# Combine both lists 
off_internal_hosts_ips_v4: "{{ (groups['all'] | map('extract', hostvars, ['ansible_ssh_host'
]) | list | reject('undefined')) + off_additional_ips_v4 }}"
```

alexgarel replied:
Great, I had it in mind, but I didn't took the time to search for it, thanks.

Additional ips should indeed be in the inventory, I will use a variable for that.

alexgarel replied:
Done.

teleboas commented on group_vars/all/common.yml:
```diff
@@ -0,0 +1,38 @@
+# FIXME
+off_private_networks_v6: []
+
+# Are we a proxmox host ?
```
From  a separation of concerns perspective,  shouldn't this be moved to `group_vars/proxmox_nodes/proxmox.yml`?

alexgarel replied:
If I move it to group_vars/proxmox_nodes/ it means it is just "true"

I can let it be undefined for other hosts (and maybe use a bool filter)

teleboas commented on roles/base/tasks/hostname.yml:
```diff
@@ -29,7 +29,7 @@
     dest: /etc/hosts
     regexp: '^127\.0\.0\.1[\s\t]+'
-    line: '127.0.0.1 {{ inventory_hostname }}.{{ host_domain }} {{ inventory_hostname }} localhost'
+    line: '127.0.0.1 {{ inventory_hostname }}.localhost localhost'
```
This change is concerning for several reasons:

  1. The original line properly mapped the server's FQDN and hostname to 127.0.0.1, while
  the new line uses a non-standard hostname pattern inventory_hostname.localhost.
  2. The modified setup could cause name resolution issues. Applications expecting to
  resolve the actual hostname to 127.0.0.1 may fail.
  3. Lines 37-56 attempt to properly map the hostname to the server's actual IP addresses,
   but removing the hostname from the localhost line could create inconsistencies.
  4. The change creates a non-standard host configuration where the FQDN doesn't follow
  the expected pattern.

  I'd recommend keeping the original configuration:
  127.0.0.1 {{ inventory_hostname }}.{{ host_domain }} {{ inventory_hostname }} localhost

  This follows standard practices for Linux server configuration.


PS:  It looks like Mailjet has been rejecting some emails sent from the address config-op@localhost. This is likely an unintended consequence of this change.

alexgarel replied:
This is causing problem to proxmox.

I don't understand why FQDN should resolve to 127.0.0.1 instead of the public address. 
This means you FQDN resolve to more than one address ?

alexgarel replied:
Ok, I imagine you resolve to 127.0.0.1 for the FQDN inside the machine (while outer world resolve to hostname).

On previous server, we used to instead resolve to cluster private ip (10.0.0.1 for example).

I will try again to use localhost and see if it works.

teleboas replied:
Both approaches can work, but using 127.0.0.1 is generally more resilient since the loopback interface is always available even when network interfaces (public or private) are down.

alexgarel replied:
Ok, I've seen that first line take precedence in /etc/hosts if we define multiple time the same name. I will at least comment on that, but rollback my change otherwise.

teleboas commented on roles/fail2ban-iptables/defaults/main.yml:
```diff
@@ -0,0 +1,14 @@
+# you might set it to INFO on service with low traffic or to monitor more closely
+fail2ban_iptables_loglevel: NOTICE
+fail2ban_iptables_ignoreips: >-
```
The fail2ban_iptables_ignoreips setting is using organization-specific variables:
  - off_internal_hosts_ips_v4
  - off_private_networks_v4
  - off_internal_hosts_ips_v6
  - off_private_networks_v6

  These variables (with the "off_" prefix) are clearly specific to Open Food Facts.

  For a truly reusable role:
  1. The role should use generic variable names
  2. It should not assume organization-specific variables exist
  3. It should provide sensible defaults that work in any environment

alexgarel replied:
I will move it in the role.

teleboas commented on roles/fail2ban-nftables/meta/main.yml:
This file should not be empty

teleboas commented on roles/iptables/templates/iptables_rules.j2:
```diff
@@ -0,0 +1,78 @@
+{% endfor %}
+# Internal host exposed ports
+{% for port in iptables_internal_hosts_only_ports %}
+-A INPUT -m set --match-set {{ internal_host_set }} src -p tcp -m tcp --dport port -j ACCEPT
```
Shouldn't it be `--dport {{ port }}` rather than `--dport port`?

alexgarel replied:
Well spotted.

teleboas reviewed (commented):
The PR introduces valuable functionality for Hetzner servers provisioning. The main
  concern is that while the user management approach with both SSH keys and passwords is  well-designed, the encryption of passwords will fail for users using non-RSA public  keys, which requires attention.

  Most other adjustments fall under these categories:

  1. Enhancing role reusability:
    a. Improve separation of concerns 
    b. Remove organization-specific variables from roles
    c. Rename the sshd role to better reflect its actual purpose
  2. Improving maintainability:
    a. Avoid duplication by using variables that can be inferred from inventory
    b. Fix the hostname configuration to follow standard practices

  The inventory structure with multiple host groups is appropriate and follows Ansible
  best practices. With these adjustments, the PR will provide a more robust and
  maintainable foundation for infrastructure management.

alexgarel committed:
fix: let user create password on first login

alexgarel committed:
refactor: make system users creation a different role

alexgarel committed:
feat: compute ip addresses from inventory

alexgarel committed:
refactor: don't use global variables in a role

alexgarel committed:
fix: iptables rule typo

alexgarel committed:
fix: get back 127.0.0.1 to map fqdn in /etc/hosts

alexgarel committed:
feat: finish fail2ban role
and add a proxmox jail

alexgarel commented:
@teleboas I have fixed the concerns you raised in your review.

I also finished the fail2ban role and added a proxmox jail.

I would like a final review.

This is not the end of the install, but I will continue in another PR.

teleboas commented on roles/fail2ban-iptables/tasks/copy_jail_files.yml:
```diff
@@ -0,0 +1,34 @@
+  local_action:
+    module: ansible.builtin.find
+    paths: "{{ role_path }}/files"
+    patterns: ".*{{ jail }}\\..*"
```
This pattern could match many unintended files if the `{{ jail }}` variable is not set, which could be risky. To ensure the playbook runs safely and as expected, it would be a good idea to initialize the `{{ jail }}` variable in `defaults/main.yml`.

alexgarel replied:
Thanks I did use a prefixed name, added a default value and check it's not empty. It was indeed a big pitfall.

teleboas reviewed (approved):
It looks great to me! Apart from a small concern about a variable initialization (see the comment below), everything else seems ready for merging. 😊

alexgarel committed:
fix: secure variable usage in copy_jail_files
from fail2ban_iptables

alexgarel committed:
Merge branch 'main' into hetzner-install
